### ISSUE TITLE
getmouselocation doesn't return window id in same form as, e.g. getactivewindow

### ISSUE THREAD

**Issue by Lew-Rockwell-Fan** [jordansissel/xdotool#191]
In Lxterminal under Openbox in a minimalist Ubuntu 16.04, 64 bit (think Lubuntu but lighter) with the mouse in the same position in the middle of the terminal:

`$ xdotool getwindowfocus
29381153
$ xdotool getactivewindow
29381153
$ xdotool getmouselocation
x:710 y:451 screen:0 window:8472033

`
Shouldn't these all return the same window id in the same format?

I'm trying to get something out of getmouselocation that I can use to query some program like xprop, obxrop, or wmctrl & find out about the window under the mouse. I can do that with the 8 digit decimal window id that getactivewindow & getwindowfocus return, but not with the 7 digit decimal number returned by getmouselocation. Is this broken or am I being stupid?

**Comment by jordansissel** [comment]
> Is this broken or am I being stupid?

For starters, you aren't being stupid -- X11, window managers, and xdotool are complex things and it's really easy for bugs to happen. My hunch is that this is a bug.

I took a look at the code, and it's not apparent to me why this happens. I haven't tried to reproduce it yet, though.

I assume you are expecting the same number to be returned by all 3 commands? 

I wonder if getmouselocation is getting stuck trying to find the top level window where the mouse cursor is.

Looking at the code:

* [we ask X11 what window the mouse is over](https://github.com/jordansissel/xdotool/blob/master/xdo.c#L885)
* [then we take this window and try and find which top-level window it belongs to](https://github.com/jordansissel/xdotool/blob/master/xdo.c#L901-L905)

Can you provide some extra details:
* What window manager are you using?
* Does this happen for only certain apps, or all app windows?

**Comment by Lew-Rockwell-Fan** [comment]
Thank you.

> What window manager are you using?
Openbox

Just to be gin clear:
No DE per se, just Openbox.
No DM - I use startx, defaulting to Openbox.
The system was built from the mini.iso up, not by tearing down from the flagship Ubuntu. EDITOR is bash. I have other WMs installed I could test with, and other interpreters ("shells") also, if you think either might be enlightening.

> Does this happen for only certain apps, or all app windows?
Happens under 2 I've tested, one a terminal, one not. I'll write something with a sleep and test some more.

> I wonder if getmouselocation is getting the wrong information while trying to find the top level window . . .
Good point. I'll experiment with situations with only 1 window & and with minimizing ("iconofying") all but the window in question. Still, it isn't just a different number, it is a different number of digits. I tried 0 padding it, but not after minimizing other windows.

> I assume you are expecting the same number to be returned by all 3 commands?
Yes, that would seem the most intuitive expectation. I use cut get just the number, that part of the format I can deal with. To some degree I can convert from decimal to hex and vice versa,. I tried that & got nownere. But there I'm confused by the fact that programs that report window id in hex don't all repert hex values with the same number of characters.

I appreciate your promptness but don't break your neck this time. I have to be away for a few hours.  Thanks. I shall return (now why did the news make me think of that line?)

**Comment by Lew-Rockwell-Fan** [comment]
tldr summary:
Testing with only 1 window open at a time (unless you count the "root window" which is just the borderless space most gui systems draw a "Desktop" on. For me, it is just a featureless field I can click on for the Openbox menus because I don't have anything set to draw a "Desktop".) using a script running in the background, gxmessage to interact with it, gmrun and the Openbox main menu to invoke programs, with appropriate sleeps to avoid reacting to the gxmessage windows or the runbox, and to give me time to center the pointer in the client area of the window, with one quasi-exception, it works the same way for all windows I tested, to whit:
lxterminal
terminator
xvt
rxvt
xterm
feh
synaptic
spacefm
thunar
convertall

In all those cases,
xdotool getwindowfocus
&
xdotool getactivewindow
report the same number (for that particular window I mean) & it is an 8 digit decimal number.
while
xdotool getmouselocation | cut -d':' -f5
reports a 7 digit decimal number not related to the other numbers in any obvious way

The only exception is when I ran a loop with NO window open. In that case getactivewindow still reported an 8 digit decimal number, but getwindowfocus reported a 7 digit decimal number and getmouselocation reported a 4 digit decimal number, none of which were related in any obvious way.

I can post a copy of the script I made to run the test and it's complete report if you want, but that is the gist of it. All done under plain Openbox with bash.

**Comment by Local9** [comment]
> If this is a bug, if anyone has a suggestion for a work around, I'd appreciate it.

I'm seeing the same problem, I am wanting to track the mouses location and know which window it is over not set the window. If there are better ways or can this bug be fixed I'd love to know.

**Comment by jordansissel** [comment]
I’ll agree this is a bug. We can probably change xdotool tonuse the same
code for determining the window id when calling getmouselocation

**Comment by rahatzamancse** [comment]
Same problem here.
I am using manjaro-bspwm. I am using the DidIClickDesktop.sh from [bspwm-scripts](https://github.com/Chrysostomus/bspwm-scripts/blob/master/bin/DidIClickDesktop.sh) and it does not work.

**Comment by frostworx** [comment]
hi, 
just stumbled over this issue and as arch linux doesn't apply any custom patches in the xdotool [PKGBUILD](https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/xdotool) it was apparently fixed at some point:

```
$ xdotool getwindowfocus
79691783
$ xdotool getactivewindow
79691783
$ xdotool getmouselocation
x:2395 y:2012 screen:0 window:79691783
```

**Comment by Lew-Rockwell-Fan** [comment]
Not fixed for me. I'm on a Pclos system now, still plain Openbox as in OP.
$ xdotool getwindowfocus
35651587
$ xdotool getactivewindow
35651587
$ xdotool getmouselocation
x:340 y:326 screen:0 window:8723655

**Comment by frostworx** [comment]
xdotool version 3.20160805.1 here

**Comment by Lew-Rockwell-Fan** [comment]
Same version here. Still doesn't work
`$ xdotool --version
xdotool version 3.20160805.1
$ xdotool getwindowfocus
73400323
$ xdotool getactivewindow
73400323
$ xdotool getmouselocation
x:327 y:329 screen:0 window:8504311
$`

I suggest a more likely hypothesis than "its been fixed" is that the bug was never manifested in Arch. That might be useful information to somebody trying to track it down.

**Comment by Neko-Box-Coder** [comment]
Hi, 
I have never had this issue until recently when I update my Arch system, where I have a script that auto focus on the window the cursor is on. 

I am running OpenBox as well with xdotool version 4.20260303.1

@Lew-Rockwell-Fan, did you ever resolve it? 
@frostworx, since I am having this issue in Arch, I wonder if you have the same issue now? (if you are still using Arch that is)

**Comment by Lew-Rockwell-Fan** [comment]
@Neko-Box-Coder  Short answer: No. I don't recall at the moment, what I was using this for, but I'm pretty sure I found some totally different way to accomplish my objective. Sorry that's kind of vague, but this was a while back. My main system now is the standard Lubuntu 24.04 but logged in to the plain Openbox environment instead of the fancier default. Testing, I see the ID numbers do have the same number of digits now, but they're not the same. Again, in LXterminal with the pointer over the terminal:
`j@2404:~$ uname -a
Linux 2404 6.17.0-14-generic #14~24.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Jan 15 15:52:10 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
j@2404:~$ wmctrl -m
Name: Openbox
Class: 
PID: N/A
Window manager's "showing the desktop" mode: OFF
j@2404:~$ xdotool --version
xdotool version 3.20160805.1
j@2404:~$ xdotool getwindowfocus
6291459
j@2404:~$ xdotool getactivewindow
6291459
j@2404:~$ xdotool getmouselocation
x:1637 y:930 screen:0 window:2373541
j@2404:~$ date
Wed Apr 15 06:31:05 PM EDT 2026
j@2404:~$ wmctrl -l
0x00800003 -1 2404 yad_toolbar-small_button_general
0x00a00004  0 2404 getmouselocation doesn't return window id in same form as, e.g. getactivewindow · Issue #191 · jordansissel/xdotool — Mozilla Firefox
0x00600003  0 2404 lxterminal
0x01c00003 -1 2404 Left_corner_yad_v2
0x01000003  0 2404 yad_toolbar_for_lxterminal
j@2404:~$ 

`

**Comment by Lew-Rockwell-Fan** [comment]
I have no clue why much of that is overstruck. I used the <> button with the hovertext "code" which wrapped what I pasted with back ticks. I can see the ticks. Anyway, I think you can still read it.

**Comment by Neko-Box-Coder** [comment]
Don't worry, I think I fixed it. I will create a PR soon.

**Comment by Lew-Rockwell-Fan** [comment]
PR? Whazzat? If you got it to work, I'd be curious how you did it.

**Comment by Lew-Rockwell-Fan** [comment]
I'm migrating to EndeavourOS which is, by my understanding, Arch with about a 2-week safety lag in incorporating upstream packages, so, but 'buntu standards, bleeding edge. I still use a plain Openbox DE, if you want to call that a DE. And this still does not work for me in that environment:
[j@ndv ~]$ xdotool getwindowfocus
27265748
[j@ndv ~]$ xdotool getactivewindow
27265748
[j@ndv ~]$ xdotool getmouselocation
x:1207 y:803 screen:0 window:4238289
[j@ndv ~]$ xdotool --version
xdotool version 4.20260303.1
[j@ndv ~]$ uname -a
Linux ndv 7.0.12-arch1-1 #1 SMP PREEMPT_DYNAMIC Wed, 10 Jun 2026 08:58:02 +0000 x86_64 GNU/Linux
[j@ndv ~]$

**Comment by Neko-Box-Coder** [comment]
@Lew-Rockwell-Fan 

Could you test my PR branch and see if it works for you?

Simply do:
```shell
mkdir /tmp/xdotoolFix && cd /tmp/xdotoolFix
git clone -b FixFindWindowClient https://github.com/Neko-Box-Coder/xdotool.git
cd xdotool && make static
./xdotool.static getwindowfocus
```

@jordansissel 

Sorry to ping you but it would be great if you can take a look and maybe merge the PR for this issue.

**Comment by Lew-Rockwell-Fan** [comment]
`Fixed` as of what version? Version 4.20260303.1 shows as the current version on the front page here and as of Sun 19 Jul 2026 that version is still broken the same way it has been in Endeavour (which is essentially Arch):

[j@ndv ~]$ xdotool getwindowfocus
31457283
[j@ndv ~]$ xdotool getactivewindow
31457283
[j@ndv ~]$ xdotool getmouselocation
x:1287 y:835 screen:0 window:4394749
[j@ndv ~]$ xdotool --version
xdotool version 4.20260303.1
[j@ndv ~]$ cat /etc/os-release |head -1
NAME="EndeavourOS"
[j@ndv ~]$ openbox --version |head -1
Openbox 3.6.1
[j@ndv ~]$ lxterminal --version
lxterminal 0.4.1
[j@ndv ~]$

**Comment by jordansissel** [comment]
@Lew-Rockwell-Fan good question! The patch #513 was merged only a few weeks ago, and there hasn't been a new release yet.

I’ll try to publish a new release in the next week or so, and I confess that most of my xdotool energy lately is going into prototypes and research to [add Wayland support](https://www.kickstarter.com/projects/jordansissel/yes-way-xdotool-for-wayland).
